packages/api/src/services/notification/notification.service.ts

```diff
@@ -0,0 +1,137 @@
+import {Service} from '../../providers/service/service';
+import {Queue} from '../../models/data-structures/queue';
+import {Notification} from '../../models/notification';
+import {service} from '../../providers';
+
+/**
+ * A service for managing notifications.
+ *
+ * This service provides methods to enqueue, dequeue, peek  notifications. It also allows subscribers to be notified
+ * whenever a new notification is added, while keeping full control over queue manipulation.
+ */
+export class NotificationService implements Service {
+
+  /**
+   * Internal list of callback functions to notify when a notification is added.
+   */
+  private notificationAddedCallbackFunctions: (() => void)[] = [];
+
+  /**
+   * Internal queue storing the notifications.
+   */
+  private readonly notifications: Queue<Notification> = new Queue<Notification>();
+
+  /**
+   * Adds a notification to the end of the notification queue.
+   *
+   * @param notification - The notification to enqueue.
+   */
+  enqueue(notification: Notification): void {
+    this.notifications.enqueue(notification);
+    this.notifyAllNotificationAdded();
+  }
+
+  /**
+   * Removes and returns the first notification from the notification queue.
+   *
+   * @returns The dequeued notification, or `undefined` if the queue is empty.
+   */
+  dequeue(): Notification | undefined {
+    return this.notifications.dequeue();
+  }
+
+  /**
+   * Removes all notifications from the notification queue and returns them.
+   *
+   * @returns An array containing all items that were in the queue.
+   */
+  dequeueAll(): Notification[] {
+    return this.notifications.dequeueAll();
+  }
+
+  /**
+   * Returns the first notification in the notification queue without removing it.
+   *
+   * @returns A deep copy of the first notification in the queue, or `undefined` if the queue is empty.
+   */
+  peek(): Notification | undefined {
+    const notification = this.notifications.peek();
+    if (notification) {
+      return notification.copy();
```

Comment:
So. You clone the notifications returned by the `peek` and `peekAll` method, but not those from `dequeue` and `dequeueAll`. Although those return removed instances, I believe we should return copies from these as well.